Introduce instruction label for memory access and generate (PTE) fault check to labels.

ShaleXIONG opened the pull request:
Introduce instruction label for memory access and generate (PTE) fault check to labels in `Diyone7`.

Most key changes are in file `gen/cycle.ml` `gen/final.ml` and `gen/AArch64Compile_gen.ml`. We extend the `node` (`cycle.ml`) to track if the associated event, e.g., memory load and store, needs to (1) check faults, and/or (2) check the effects, (for example the loaded value). This generate better post-conditions.
- If there is PTE value change in the cycle, the next memory access will be labelled and check fault or not fault, and
- If there is a (plain) value change, the next memory load will checked the loaded value (Consistent with the previous behaviour but rule out redundant register value check in the litmus tests purely for fault check).

 Minor code quality improvement:
- Replace a few manually fold and iteration over lists by standard library call of `List.fold_left`, `List.iter`, `List.map` etc. This are most in `gen/top_gen.ml` and `gen/alt.ml`.
- Add a few comments in some key functions purely for readability.

ShaleXIONG commented:
> Hi @ShaleXIONG,
> 
> This look really interesting! Do you think that idea of taking faults could be also applied to MTE (aka memtag)?

I am not expert on how MTE works. However, I have read most the code in `diyone7` and I am aware the code related to MTE. My instinct is, I can do similar to MTE but I will need some small examples from your side.

relokin commented:
Thanks Shale, is this ready for review? @hugookeeffe can you have a look as soon as Shale mark's this as ready?

ShaleXIONG commented:
> Thanks Shale, is this ready for review? @hugookeeffe can you have a look as soon as Shale mark's this as ready?

It is ready, but maybe let @hugookeeffe review first. Once he is happy on the change, I will squash all the commit into one, and add more description such as put some examples etc.

relokin commented:
> > Thanks Shale, is this ready for review? @hugookeeffe can you have a look as soon as Shale mark's this as ready?
> 
> It is ready, but maybe let @hugookeeffe review first. Once he is happy on the change, I will squash all the commit into one, and add more description such as put some examples etc.

Super thanks Shale.

hugookeeffe commented on gen/cycle.ml:
```diff
@@ -817,136 +871,163 @@ let set_same_loc st n0 =
+                       let next_loc () =
+                         match n.evt.loc with
+                         | Code.Data x ->
+                            begin try
```
Would it be worth putting this in a seperate function to improve readability in this function?

ShaleXIONG replied:
I did not change this part, though I will have a look what I can do.

ShaleXIONG replied:
I will do this as a separate merge, once I find a solution to make it more readable.

hugookeeffe commented on gen/edge.ml:
```diff
@@ -583,7 +583,7 @@ let fold_tedges f r =
 
 
-  let t = Hashtbl.create 101
+  let t = Hashtbl.create 40000
```
What is the prupose of this change? Does it affect the hash of created litmus tests?

ShaleXIONG replied:
No it does not affect the functionality. This is the initial size of the hashtable. I notice there are a lot of entry in the hashtable in `AArch64` situation so I change to a larger initial size.

hugookeeffe commented on gen/final.ml:
```diff
@@ -189,9 +204,13 @@ module Make : functor (O:Config) -> functor (C:ArchRun.S) ->
            assert (evt.C.C.bank = Code.Ord || evt.C.C.bank = Code.CapaSeal) ;
            Some (I (prev_value evt.C.C.v))
         | None|Some Code.J -> None in
+        (* *)
```
Should this be in?

ShaleXIONG replied:
Removed.

hugookeeffe commented on gen/top_gen.ml:
```diff
@@ -168,11 +167,11 @@ module U = TopUtils.Make(O)(Comp)
             Comp.emit_load_one st p init loc in
           Some r,init,i,st
-        else
+        else begin
```
is it necessary to add this begin ... end catch?

ShaleXIONG replied:
Removed.

hugookeeffe reviewed (commented):
Thanks for the PR! Overall the changes look good, thanks for this!

I have made a variety of comments, but most of them are nitpicks.

I have a couple of overall comments:
- As the PR is quite big, i think it would be worth putting the general code improvements and syntax improvements into a different PR where possible. If this is too much work then its not necessary, but it is diffuclt to identify what changes are important to look at at times.
- What testing has been done?
- I think it would be valuable to give examples in the description of the PR of what litmus tests this allows us to generate.
- It might be useful to have an explanation or description of the major changes to the code to describe why they have been (i.e. additions to event structure and changes to CoST)

ShaleXIONG commented on gen/AArch64Compile_gen.ml:
```diff
@@ -1777,7 +1783,7 @@ module Make(Cfg:Config) : XXXCompile_gen.S =
             Some r,init,cs@pseudo cs2,st
         | R,Some (AcqPc _,None) ->
-            let r,init,cs,st = LDAPR.emit_load st p init loc  in
+            let r,init,cs,st = LDAPR.emit_load st p init loc in
```
I remove an extra space here.

ShaleXIONG commented on gen/FaultSet.ml:
```diff
@@ -0,0 +1,17 @@
+(* "http://www.cecill.info". We also give a copy in LICENSE.txt.            *)
+(****************************************************************************)
+
+include MySet.Make(FaultSetElm)
```
There are two places use it, `final.ml` and `top_gen.ml`. I am not expert on the module system, do you think it is better I put in for example `final.ml` and import the `FaultSet` to `top_gen.ml`?

hugookeeffe replied:
to me its mostly a matter of preference I think. you could put it in final and import it i believe yes. It looks like top_gen has access to the final module so should be possible

ShaleXIONG replied:
I move the FaultSet definition into `final.ml`.

ShaleXIONG commented on gen/alt.ml:
```diff
@@ -685,12 +660,10 @@ module Make(C:Builder.S)
-            (zyva_prefix rem aset relax safe reject n f k )
+    (* TODO: explain the parameter here.
+       Parameter `f` is a function check the result? *)
+    let zyva_prefix prefixes aset relax safe reject n f k =
```
That is a good idea. I will try to do some splits.

ShaleXIONG commented on gen/AArch64Compile_gen.ml:
```diff
@@ -1926,7 +1932,22 @@ module Make(Cfg:Config) : XXXCompile_gen.S =
+        end in
+        (* Add the instruction label to `cs`,
+           when the corresponding node requires a fault check. *)
+        let cs = match e.C.check_fault with
```
it is a good idea. Will do.

ShaleXIONG commented on gen/AArch64Arch_gen.ml:
```diff
@@ -584,10 +585,16 @@ let overwrite_value v ao w = match ao with
           | OA -> { p with oa=OutputAddress.PHY (loc ()); })
       fs p
 
+    (* function name is confused, it sets the pteval based on atom `a` *)
```
Will remove. I was a bit confused previously.

ShaleXIONG commented on gen/AArch64Arch_gen.ml:
```diff
@@ -584,10 +585,16 @@ let overwrite_value v ao w = match ao with
       | Pte f,None -> do_setpteval a f p
       | _ -> Warn.user_error "Atom %s is not a pteval write" (pp_atom a)
+
+    open AArch64PteVal
```
Done.

ShaleXIONG commented on gen/AArch64Compile_gen.ml:
```diff
@@ -1756,12 +1756,18 @@ module Make(Cfg:Config) : XXXCompile_gen.S =
             Some (a,Some (MachSize.S128,0))
           | _ -> Some (a,m) end in
-        begin match d,atom with
+        (* Compute the base value for
```
I have change to "compile the node".

ShaleXIONG commented on gen/alt.ml:
```diff
@@ -305,6 +305,9 @@ module Make(C:Builder.S)
         end
 
+    (* List.is_empty only supports for ocaml 5.1 afterwards *)
+    let is_empty_list l = List.compare_length_with l 0 = 0
```
Yes it will work. I also move this change to an separate merge (will make it soon).

ShaleXIONG commented:
@maranget @murzinv @relokin  Can you have a look on this pull? Between Hugo and I, we are happy on the change.

maranget commented on gen/AArch64Compile_gen.ml:
```diff
@@ -1725,7 +1725,20 @@ module Make(Cfg:Config) : XXXCompile_gen.S =
+      | Some (label_name, _) ->
+        (* Always label the last instruction,
+           which should be the actual load or store. *)
+        let length = List.length cs in
+        List.mapi
+        ( fun index instr ->
+            if index = (length - 1) then
+                Label(label_name, instr)
+            else instr ) cs
```
Some may find the following code to be more straightforward:
```suggestion
          let rec do_rec = function
            | [] -> assert false
            | [instr] -> [Label (label_name,instr)]
            | instr::rem -> instr::do_rec rem in
          do_rec cs
```

ShaleXIONG replied:
Will do. I will probably return [] on [] case.

ShaleXIONG commented on gen/FaultSetElm.ml:
```diff
@@ -0,0 +1,27 @@
+(****************************************************************************)
```
I managed to move the entire definition into `final.ml` since it is the only place it is used.

relokin commented on gen/top_gen.ml:
```diff
@@ -199,11 +196,15 @@ let get_fence n =
 
-  let rec compile_proc pref chk loc_writes st p ro_prev init ns = match ns with
+  (* `ro_prev` if there is a previous load, `st` is the machine state
+     The callee from compile_proc
+                  Misc.identity false loc_writes A.st0 proc No      i    n *)
```
I am not a big fun of comments like this. Either pref is always Misc.identity and we should get rid of the argument or the comment is wrong.

ShaleXIONG replied:
Will rewrite something more descriptive.

ShaleXIONG commented on gen/final.ml:
```diff
@@ -22,6 +22,22 @@ module type Config = sig
+    | label, _ -> label
+end
+
+module FaultSet = MySet.Make(FaultSetElm)
```
After some thought and experimenting hacks, I do not think reusing `lib/fault.mli` will work for now. The main blocker is `fault.mli` requires every fault contain a (non-optional) `Label.t`. Despite we now have Label for PTE but there is no label for existing MTE and morello variants. I think the best way to progress here is I will rework this part in #1254, where I extend label to MTE and I can extend label to morello too.

ShaleXIONG commented:
@relokin @maranget  I have addressed all the comments. Also split some of the code improvement change to a separate pull request (will raise the request soon).

relokin commented:
Can you please squash all commits into one and rebase? I will merge this soon after. It would be great you could add more info on how you tested either in the commit message or in the message of the PR.

ShaleXIONG commented:
> Same as for #1254 , it seems load exclusives and store exclusives need extra care. For instance
> 
> ```
> diyone7 -arch AArch64 -variant kvm "PodWWPPteVA Rfe LxSx PodWW DMB.SYdWR Fre"
> AArch64 A
> "PodWWPPteVA RfePteVAP LxSx PodWW DMB.SYdWR Fre"
> Generator=diyone.exe (version 7.58+1)
> Prefetch=0:x=F,0:y=W,1:y=F,1:x=T
> Com=Rf Fr
> Orig=PodWWPPteVA RfePteVAP LxSx PodWW DMB.SYdWR Fre
> { int x=0; int y=4; int z=8;
> 0:X1=x; 0:X2=PTE(y); 0:X3=(oa:PA(y), valid:0);
> 1:X0=y; 1:X5=z; 1:X6=x;
> }
>  P0          | P1              ;
>  MOV W0,#1   | MOV W2,#5       ;
>  STR W0,[X1] | L00:            ;
>  STR X3,[X2] | Loop00:         ;
>              | LDXR W1,[X0]    ;
>              | STXR W3,W2,[X0] ;
>              | CBNZ W3,Loop00  ;
>              | MOV W4,#9       ;
>              | STR W4,[X5]     ;
>              | DMB SY          ;
>              | LDR W7,[X6]     ;
> exists (1:X1=0 /\ 1:X7=0) /\ fault (P1:L00,y)
> ```
> 
> Doesn't account that `STXR` may fault, here is slightly altered tests
> 
> ```
> AArch64 A
> "PodWWPPteVA RfePteVAP LxSx PodWW DMB.SYdWR Fre"
> Generator=diyone.exe (version 7.58+1)
> Prefetch=0:x=F,0:y=W,1:y=F,1:x=T
> Com=Rf Fr
> Orig=PodWWPPteVA RfePteVAP LxSx PodWW DMB.SYdWR Fre
> { int x=0; int y=4; int z=8;
> 0:X1=x; 0:X2=PTE(y); 0:X3=(oa:PA(y), valid:0);
> 1:X0=y; 1:X5=z; 1:X6=x;
> }
>  P0          | P1              ;
>  MOV W0,#1   | MOV W2,#5       ;
>  STR W0,[X1] | L00:            ;
>  STR X3,[X2] | Loop00:         ;
>              | LDXR W1,[X0]    ;
> 	     | L01:            ; (* <-------- label for STXR *)
>              | STXR W3,W2,[X0] ;
>              | CBNZ W3,Loop00  ;
>              | MOV W4,#9       ;
>              | STR W4,[X5]     ;
>              | DMB SY          ;
>              | LDR W7,[X6]     ;
> exists (1:X1=0 /\ 1:X7=0) /\ (fault (P1:L00,y) \/ fault (P1:L01,y)) (* <----- updated with L01 *)
> ```
> 
> and `herd7` output
> 
> ```
>  herd7 -variant vmsa  -unroll 1 T1.litmus
> Test A Allowed
> States 5
> 1:X1=0; 1:X7=0; Fault(P1:{L00,Loop00},y,MMU:Translation); ~Fault(P1:L01,y);
> 1:X1=4; 1:X7=0;  ~Fault(P1:L01,y); ~Fault(P1:L00,y);
> 1:X1=4; 1:X7=0; Fault(P1:{L00,Loop00},y,MMU:Translation); ~Fault(P1:L01,y);
> 1:X1=4; 1:X7=0; Fault(P1:L01,y,MMU:Translation); ~Fault(P1:L00,y); <-----  note, there is state where LDXR did not fault but STXR does
> 1:X1=4; 1:X7=1;  ~Fault(P1:L01,y); ~Fault(P1:L00,y);
> Loop Ok
> Witnesses
> Positive: 1 Negative: 7
> Condition exists (1:X1=0 /\ 1:X7=0 /\ (fault(P1:L00,y) \/ fault(P1:L01,y)))
> Observation A Sometimes 1 7
> Time A 1.30
> Hash=32ea87780e96d479dc099cbf81d07a01
> 
> Warning: File "T1.litmus": unrolling limit exceeded at L00, legal outcomes may be missing.
> ```
> 
> Also, need to pay attention to `-unrollatomic <n> unroll atomic idioms (default, use loops)`, now final condition takes care only for the first iteration
> 
> ```
> $diyone7 -arch AArch64 -variant kvm "PodWWPPteVA Rfe LxSx PodWW DMB.SYdWR Fre" -ua 5
> AArch64 A
> "PodWWPPteVA RfePteVAP LxSx PodWW DMB.SYdWR Fre"
> Generator=diyone.exe (version 7.58+1)
> Prefetch=0:x=F,0:y=W,1:y=F,1:x=T
> Com=Rf Fr
> Orig=PodWWPPteVA RfePteVAP LxSx PodWW DMB.SYdWR Fre
> { int x=0; int y=4; int z=8;
> 0:X1=x; 0:X2=PTE(y); 0:X3=(oa:PA(y), valid:0);
> 1:X0=y; 1:X6=z; 1:X7=x; 1:X9=ok1;
> int ok1=0;
> }
>  P0          | P1              ;
>  MOV W0,#1   | MOV W2,#5       ;
>  STR W0,[X1] | L00:            ;
>  STR X3,[X2] | LDXR W1,[X0]    ;
>              | STXR W3,W2,[X0] ;
>              | CBZ W3,Go00     ;
>              | LDXR W1,[X0]    ;
>              | STXR W3,W2,[X0] ;
>              | CBZ W3,Go00     ;
>              | LDXR W1,[X0]    ;
>              | STXR W3,W2,[X0] ;
>              | CBZ W3,Go00     ;
>              | LDXR W1,[X0]    ;
>              | STXR W3,W2,[X0] ;
>              | CBZ W3,Go00     ;
>              | LDXR W1,[X0]    ;
>              | STXR W3,W2,[X0] ;
>              | CBNZ W3,End1    ;
>              | Go00:           ;
>              | ADD W4,W4,#1    ;
>              | MOV W5,#9       ;
>              | STR W5,[X6]     ;
>              | DMB SY          ;
>              | LDR W8,[X7]     ;
>              | End1:           ;
>              | STR W4,[X9]     ;
> exists (1:X1=0 /\ 1:X8=0 /\ [ok1]=1) /\ fault (P1:L00,y)
> ```



> Same as for #1254 , it seems load exclusives and store exclusives need extra care. For instance
> 
> ```
> diyone7 -arch AArch64 -variant kvm "PodWWPPteVA Rfe LxSx PodWW DMB.SYdWR Fre"
> AArch64 A
> "PodWWPPteVA RfePteVAP LxSx PodWW DMB.SYdWR Fre"
> Generator=diyone.exe (version 7.58+1)
> Prefetch=0:x=F,0:y=W,1:y=F,1:x=T
> Com=Rf Fr
> Orig=PodWWPPteVA RfePteVAP LxSx PodWW DMB.SYdWR Fre
> { int x=0; int y=4; int z=8;
> 0:X1=x; 0:X2=PTE(y); 0:X3=(oa:PA(y), valid:0);
> 1:X0=y; 1:X5=z; 1:X6=x;
> }
>  P0          | P1              ;
>  MOV W0,#1   | MOV W2,#5       ;
>  STR W0,[X1] | L00:            ;
>  STR X3,[X2] | Loop00:         ;
>              | LDXR W1,[X0]    ;
>              | STXR W3,W2,[X0] ;
>              | CBNZ W3,Loop00  ;
>              | MOV W4,#9       ;
>              | STR W4,[X5]     ;
>              | DMB SY          ;
>              | LDR W7,[X6]     ;
> exists (1:X1=0 /\ 1:X7=0) /\ fault (P1:L00,y)
> ```
> 
> Doesn't account that `STXR` may fault, here is slightly altered tests
> 
> ```
> AArch64 A
> "PodWWPPteVA RfePteVAP LxSx PodWW DMB.SYdWR Fre"
> Generator=diyone.exe (version 7.58+1)
> Prefetch=0:x=F,0:y=W,1:y=F,1:x=T
> Com=Rf Fr
> Orig=PodWWPPteVA RfePteVAP LxSx PodWW DMB.SYdWR Fre
> { int x=0; int y=4; int z=8;
> 0:X1=x; 0:X2=PTE(y); 0:X3=(oa:PA(y), valid:0);
> 1:X0=y; 1:X5=z; 1:X6=x;
> }
>  P0          | P1              ;
>  MOV W0,#1   | MOV W2,#5       ;
>  STR W0,[X1] | L00:            ;
>  STR X3,[X2] | Loop00:         ;
>              | LDXR W1,[X0]    ;
> 	     | L01:            ; (* <-------- label for STXR *)
>              | STXR W3,W2,[X0] ;
>              | CBNZ W3,Loop00  ;
>              | MOV W4,#9       ;
>              | STR W4,[X5]     ;
>              | DMB SY          ;
>              | LDR W7,[X6]     ;
> exists (1:X1=0 /\ 1:X7=0) /\ (fault (P1:L00,y) \/ fault (P1:L01,y)) (* <----- updated with L01 *)
> ```
> 
> and `herd7` output
> 
> ```
>  herd7 -variant vmsa  -unroll 1 T1.litmus
> Test A Allowed
> States 5
> 1:X1=0; 1:X7=0; Fault(P1:{L00,Loop00},y,MMU:Translation); ~Fault(P1:L01,y);
> 1:X1=4; 1:X7=0;  ~Fault(P1:L01,y); ~Fault(P1:L00,y);
> 1:X1=4; 1:X7=0; Fault(P1:{L00,Loop00},y,MMU:Translation); ~Fault(P1:L01,y);
> 1:X1=4; 1:X7=0; Fault(P1:L01,y,MMU:Translation); ~Fault(P1:L00,y); <-----  note, there is state where LDXR did not fault but STXR does
> 1:X1=4; 1:X7=1;  ~Fault(P1:L01,y); ~Fault(P1:L00,y);
> Loop Ok
> Witnesses
> Positive: 1 Negative: 7
> Condition exists (1:X1=0 /\ 1:X7=0 /\ (fault(P1:L00,y) \/ fault(P1:L01,y)))
> Observation A Sometimes 1 7
> Time A 1.30
> Hash=32ea87780e96d479dc099cbf81d07a01
> 
> Warning: File "T1.litmus": unrolling limit exceeded at L00, legal outcomes may be missing.
> ```
> 
> Also, need to pay attention to `-unrollatomic <n> unroll atomic idioms (default, use loops)`, now final condition takes care only for the first iteration
> 
> ```
> $diyone7 -arch AArch64 -variant kvm "PodWWPPteVA Rfe LxSx PodWW DMB.SYdWR Fre" -ua 5
> AArch64 A
> "PodWWPPteVA RfePteVAP LxSx PodWW DMB.SYdWR Fre"
> Generator=diyone.exe (version 7.58+1)
> Prefetch=0:x=F,0:y=W,1:y=F,1:x=T
> Com=Rf Fr
> Orig=PodWWPPteVA RfePteVAP LxSx PodWW DMB.SYdWR Fre
> { int x=0; int y=4; int z=8;
> 0:X1=x; 0:X2=PTE(y); 0:X3=(oa:PA(y), valid:0);
> 1:X0=y; 1:X6=z; 1:X7=x; 1:X9=ok1;
> int ok1=0;
> }
>  P0          | P1              ;
>  MOV W0,#1   | MOV W2,#5       ;
>  STR W0,[X1] | L00:            ;
>  STR X3,[X2] | LDXR W1,[X0]    ;
>              | STXR W3,W2,[X0] ;
>              | CBZ W3,Go00     ;
>              | LDXR W1,[X0]    ;
>              | STXR W3,W2,[X0] ;
>              | CBZ W3,Go00     ;
>              | LDXR W1,[X0]    ;
>              | STXR W3,W2,[X0] ;
>              | CBZ W3,Go00     ;
>              | LDXR W1,[X0]    ;
>              | STXR W3,W2,[X0] ;
>              | CBZ W3,Go00     ;
>              | LDXR W1,[X0]    ;
>              | STXR W3,W2,[X0] ;
>              | CBNZ W3,End1    ;
>              | Go00:           ;
>              | ADD W4,W4,#1    ;
>              | MOV W5,#9       ;
>              | STR W5,[X6]     ;
>              | DMB SY          ;
>              | LDR W8,[X7]     ;
>              | End1:           ;
>              | STR W4,[X9]     ;
> exists (1:X1=0 /\ 1:X8=0 /\ [ok1]=1) /\ fault (P1:L00,y)
> ```

Will investigate. I will cherry pick the label fix for `RMW` instruction from the #1254 and find a special strategy for `LxSx` since it needs two label with logical or in the final exist check.

relokin commented:
@murzinv, I am not sure I see why this relaxation list `PodWWPPteVA RfePteVAP LxSx PodWW DMB.SYdWR Fre` produces an interesting test. But besides that, I also don't see why it should produce this test:
```
AArch64 A
"PodWWPPteVA RfePteVAP LxSx PodWW DMB.SYdWR Fre"
Generator=diyone7 (version 7.58+1)
Prefetch=0:x=F,0:y=W,1:y=F,1:x=T
Com=Rf Fr
Orig=PodWWPPteVA RfePteVAP LxSx PodWW DMB.SYdWR Fre
{ int z=0; int y=4; int x=8;
0:X1=x; 0:X2=PTE(y); 0:X3=(oa:PA(y), valid:0);
1:X0=y; 1:X8=z; 1:X1=x;
}
 P0          | P1              ;
 MOV W0,#9   | MOV W5,#5       ;
 STR W0,[X1] | CBNZ W6,Loop00  ;
 STR X3,[X2] | Loop00:         ;
             | L01:            ;
             | LDXR W4,[X0]    ;
             | L00:            ;
             | STXR W6,W5,[X0] ;
             | MOV W7,#1       ;
             | STR W7,[X8]     ;
             | DMB SY          ;
             | LDR W9,[X1]     ;
exists (1:X4=4 /\ 1:X9=8) /\ (Fault(P1:L00,y) \/ Fault(P1:L01,y))
```
This test is not what the relaxation list is asking for. Notice the Rfe in the prefix ``PodWWPPteVA RfePteVAP LxSx ...`. That edge means that the Implicit TTD Memory Read Effect for the LDXR is reading from the Explicit Memory Write Effect and the execution where the STXR faults means that the LDXR didn't fault.

relokin commented:
IMO, this test `PodWWPPteVA Rfe LxSx PodWW DMB.SYdWR Fre` *should not check* for a fault in the STXR because that would imply that the LDXR didn't fault and that contradicts the Rfe relaxation and we would need a Fre instead. My suggestion would be to remove the 2nd commit from this PR.

ShaleXIONG commented on gen/archExtra_gen.ml:
```diff
@@ -62,6 +62,7 @@ module type S = sig
 
 (* complete init with necessary information *)
   val complete_init : bool (* hexa *) -> Code.env -> init -> init
+  val pp_env: init -> string
```
Solved in a new commit, (will rebase to previous once settled). It is a debug pretty print that was use in `cycle.ml` but the actual code in `cycle.ml` is often commented out.
